src/app/reporting/SynchronizedReportSchedulerImpl.cpp

```diff
@@ -0,0 +1,178 @@
+    }
+}
+
+/// @brief Checks if the timer is active for the given ReadHandler. Since all read handlers are scheduled on the same timer, we
+/// check if the node is in the list and if the timer is active for the ReportScheduler
+bool SynchronizedReportSchedulerImpl::IsReportScheduled(ReadHandler * aReadHandler)
+{
+    ReadHandlerNode * node = FindReadHandlerNode(aReadHandler);
+    VerifyOrReturnValue(nullptr != node, false);
+    return mTimerDelegate->IsTimerActive(this);
+}
+
+/// @brief Find the smallest maximum interval possible and set it as the common maximum
+/// @return NO_ERROR if the smallest maximum interval was found, error otherwise, INVALID LIST LENGTH if the list is empty
+CHIP_ERROR SynchronizedReportSchedulerImpl::FindNextMaxInterval()
+{
+    VerifyOrReturnError(!mReadHandlerList.Empty(), CHIP_ERROR_INVALID_LIST_LENGTH);
+    System::Clock::Timestamp now      = mTimerDelegate->GetCurrentMonotonicTimestamp();
+    System::Clock::Timestamp earliest = now + Seconds16(kSubscriptionMaxIntervalPublisherLimit);
+
+    for (auto & iter : mReadHandlerList)
+    {
+        if (iter.GetMaxTimestamp() < earliest && iter.GetMaxTimestamp() > now)
+        {
+            earliest = iter.GetMaxTimestamp();
+        }
+    }
+
+    mNextMaxTimestamp = earliest;
+
+    return CHIP_NO_ERROR;
+}
+
+/// @brief Find the highest minimum timestamp possible that still respects the lowest max timestamp and sets it as the common
+/// minimum. If the max timestamp has not been updated and is in the past, or if no min timestamp is lower than the current max
+/// timestamp, this will set now as the common minimum timestamp, thus allowing the report to be sent immediately.
+/// @return NO_ERROR if the highest minimum timestamp was found, error otherwise, INVALID LIST LENGTH if the list is empty
+CHIP_ERROR SynchronizedReportSchedulerImpl::FindNextMinInterval()
+{
+    VerifyOrReturnError(!mReadHandlerList.Empty(), CHIP_ERROR_INVALID_LIST_LENGTH);
+    System::Clock::Timestamp latest = mTimerDelegate->GetCurrentMonotonicTimestamp();
+
+    for (auto & iter : mReadHandlerList)
+    {
+        if (iter.GetMinTimestamp() > latest)
+        {
+            // We do not want the new min to be set above the max for any handler
+            if (iter.GetMinTimestamp() <= mNextMaxTimestamp)
+            {
+                latest = iter.GetMinTimestamp();
+            }
+        }
+    }
+
+    mNextMinTimestamp = latest;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SynchronizedReportSchedulerImpl::CalculateNextReportTimeout(Timeout & timeout, ReadHandlerNode * aNode)
+{
+    VerifyOrReturnError(mReadHandlerList.Contains(aNode), CHIP_ERROR_INVALID_ARGUMENT);
+    ReturnErrorOnFailure(FindNextMaxInterval());
+    ReturnErrorOnFailure(FindNextMinInterval());
+
+    Timestamp now = mTimerDelegate->GetCurrentMonotonicTimestamp();
+    // Find out if any handler is reportable now
+    if (IsReadHandlerReportable(aNode->GetReadHandler()))
```

Comment:
The comment was not properly updated. We are just checking the current handler since we will likely end up here when several reports are scheduled. If we did check all of them, we could end up in a loop because handler1 see that handler2 is reportable so tries to sink with it. We only we to sink the other handler if the current handler is reportable

Reply:
I'm not sure I follow.  Say we call CalculateNextReportTimeout for handlerA becase it just finished the priming read for its subscription, which calls into this code.

At this point handlerA is not reportable (it just finished reporting).  So we will set `timeout` so that our next sync timestamp is `mNextMaxTimestamp`.

But what if we already have a handlerB sitting around, which is reportable, and was just waiting for its min-inteval to elapse?

In this situation, are we not just going to fail to report the things that should have been reported on handlerB?

Reply:
> With a common min timestamp handler1 would already be scheduled to wake up on handler2's min

I don't see why that matters.  It doesn't matter what the mins are at _all_ in the scenario I described.  Because the incoming handler is not reportable, we only look at the max values, no?  Please note: it is _not_ the same as the scenario you describe.

> it would then see handler1 as reportable

That seems extremely broken, and suggests that something is off about the whole setup.  We should not be setting sync values to 0 on things that just got reported....

It sounds like we need to really carefully think about these "intermediate" syncing updates.  Ideally, we would not have them at all...

> In the event where handler1 must emit before handler2, that means that handler2's min was higher than handler1's max (otherwise it would end up being emitted during the run right after handler1 is done).

I am not following this part.  This seems to be making some assumptions about what "must" and "emit" mean hat are very unclear to me....  Could you please unpack this part a little more?

> then we need to schedule the next wakeup to the earliest between handler1's min and the common min

But the common min is the "maximum value of the mins" right now... so why are we taking the minimum of it and the handler's min here?

I would really like us to get to a point where what we have is pretty obviously, or at least provably correct, and so far we are not there.  :(

Reply:
Appologies, it seems I assumed that both handler1 and handler2 were dirty in my test described above.

> In the event where handler1 must emit before handler2, that means that handler2's min was higher than handler1's max (otherwise it would end up being emitted during the run right after handler1 is done).

What I meant here is that normally, when setting the next timeout to handler1's max, we would have also set handler2's sync timestamp to that time, thus allowing handler2 to emit a report on the next engine run because when the engine will check through the handler pool, handler2->IsReportableNow() will give `true` unless it's min timestamp is higher than this timestamp. 

> But the common min is the "maximum value of the mins" right now... so why are we taking the minimum of it and the handler's min here?

That is not what the code is doing here, that is a compromise for better responsiveness. I am sorry, it seems I am getting mixed up between two conversation of two different comments.

Also, it seems some part of my update did not go through in my last commit, will update it shortly.